wgengine/magicsock/magicsock_linux.go

```diff
@@ -82,25 +106,24 @@ var (
 	// and thus we'd rather be conservative here and possibly not receive
 	// disco packets rather than slow down the system.
 	magicsockFilterV6 = []bpf.Instruction{
-		// For raw UDPv6 sockets, BPF receives _only_ the UDP header onwards, not an entire IP packet.
-		//
-		//    https://stackoverflow.com/questions/24514333/using-bpf-with-sock-dgram-on-linux-machine
-		//    https://blog.cloudflare.com/epbf_sockets_hop_distance/
-		//
-		// This is especially confusing because this *isn't* true for
-		// IPv4; see the following code from the 'ping' utility that
-		// corroborates this:
-		//
-		//    https://github.com/iputils/iputils/blob/1ab5fa/ping/ping.c#L1667-L1676
-		//    https://github.com/iputils/iputils/blob/1ab5fa/ping/ping6_common.c#L933-L941
+		// Do a bounds check to ensure we have enough space for a disco
+		// packet; see the comment in the IPv4 BPF program for more
+		// details.
+		bpf.LoadAbsolute{Off: uint32(ipv6.HeaderLen + udpHeaderSize + discoMinHeaderSize - 1), Size: 1},
```

Comment:
In this filter, the reasoning of the IPv4 version doesn't apply, because IPv6 has no bespoke LoadMemShift instruction, if you want to skip extensions you have to walk the linked list of next protocols by hand, and because v6 extensions are rare in practice, the tests below just bail on extension-ful packets. So, in this instance, testing the payload size upfront doesn't change the set of packets we accept/reject here.